consul/resource_consul_configuration_entry.go

```diff
@@ -0,0 +1,176 @@
+	return &schema.Resource{
+		Create: resourceConsulConfigurationEntryCreate,
+		Update: resourceConsulConfigurationEntryCreate,
+		Read:   resourceConsulConfigurationEntryRead,
+		Delete: resourceConsulConfigurationEntryDelete,
+
+		Schema: map[string]*schema.Schema{
+			"kind": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				ValidateFunc: validation.StringInSlice(
+					[]string{
+						consulConfigEntryServiceDefaults,
+						consulConfigEntryProxyDefaults,
+					},
+					false,
+				),
+			},
+
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"protocol": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"config": {
+				Type:     schema.TypeMap,
+				Optional: true,
+			},
+
+			"token": {
+				Type:      schema.TypeString,
+				Optional:  true,
+				Sensitive: true,
+			},
+		},
+	}
+}
+
+func resourceConsulConfigurationEntryCreate(d *schema.ResourceData, meta interface{}) error {
+	return resourceConsulConfigurationEntryUpdate(d, meta)
+}
+
+func resourceConsulConfigurationEntryUpdate(d *schema.ResourceData, meta interface{}) error {
+	if err := validateConfigEntry(d); err != nil {
+		return err
+	}
+
+	var config consulapi.ConfigEntry
+	configEntries := getClient(meta).ConfigEntries()
+	configKind := d.Get("kind").(string)
+	configName := d.Get("name").(string)
+
+	switch configKind {
+	case consulConfigEntryServiceDefaults:
+		config = &consulapi.ServiceConfigEntry{
+			Kind:     consulConfigEntryServiceDefaults,
+			Name:     configName,
+			Protocol: d.Get("protocol").(string),
+		}
+	case consulConfigEntryProxyDefaults:
+		config = &consulapi.ProxyConfigEntry{
+			Kind:   consulConfigEntryProxyDefaults,
+			Name:   configName,
+			Config: d.Get("config").(map[string]interface{}),
+		}
+	default:
+		return fmt.Errorf("Config kind '%s' is not supported", configKind)
+	}
+
+	wOpts := &consulapi.WriteOptions{
+		Token: d.Get("token").(string),
+	}
+	if _, _, err := configEntries.Set(config, wOpts); err != nil {
+		return fmt.Errorf("Failed to set '%s' config entry: %#v", configName, err)
+	}
+
+	return resourceConsulConfigurationEntryRead(d, meta)
+}
+
+func resourceConsulConfigurationEntryRead(d *schema.ResourceData, meta interface{}) error {
+	configEntries := getClient(meta).ConfigEntries()
+	configKind := d.Get("kind").(string)
+	configName := d.Get("name").(string)
+
+	qOpts := &consulapi.QueryOptions{
+		Token: d.Get("token").(string),
+	}
+	configEntry, _, err := configEntries.Get(configKind, configName, qOpts)
+	if err != nil {
+		if strings.Contains(err.Error(), "Unexpected response code: 404") {
+			// The config entry has been removed
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("Failed to fetch '%s' config entry: %#v", configName, err)
+	}
+
+	switch configKind {
+	case consulConfigEntryProxyDefaults:
+		if err = d.Set("config", configEntry.(*consulapi.ProxyConfigEntry).Config); err != nil {
```

Comment:
I'd be tempted to type assert this with, `entry, ok := configEntry.(*consulapi.ProxyConfigEntry)` and return an error on bad type rather than panic and crash. It is a bug for sure but panic and crash is not a great UX if we ever do return something bad or the response is garbled by a proxy or similar.